zerver/migrations/0423_fix_attachment_owner.py

```diff
@@ -0,0 +1,172 @@
+from datetime import timedelta
+
+from django.conf import settings
+from django.db import connection, migrations
+from django.db.backends.postgresql.schema import BaseDatabaseSchemaEditor
+from django.db.migrations.state import StateApps
+from psycopg2.sql import SQL, Identifier, Literal
+
+
+def fix_attachment_owner(apps: StateApps, schema_editor: BaseDatabaseSchemaEditor) -> None:
+    Realm = apps.get_model("zerver", "Realm")
+    UserProfile = apps.get_model("zerver", "UserProfile")
+    Client = apps.get_model("zerver", "Client")
+    Message = apps.get_model("zerver", "Message")
+    ArchivedMessage = apps.get_model("zerver", "ArchivedMessage")
+    Stream = apps.get_model("zerver", "Stream")
+    Attachment = apps.get_model("zerver", "Attachment")
+    ArchivedAttachment = apps.get_model("zerver", "ArchivedAttachment")
+
+    if not Realm.objects.exists():
+        return
+
+    mail_gateway_bot = UserProfile.objects.get(email__iexact=settings.EMAIL_GATEWAY_BOT)
+
+    # "Internal" is the client-id of all mail gateway posts
+    internal_client, _ = Client.objects.get_or_create(name="Internal")
+
+    # We only look in Attachment and not ArchivedAttachment because,
+    # never having been associated with a message, there is no way for
+    # the attachments to have been archived.
+    orphan_attachments = Attachment.objects.filter(
+        messages=None,
+        owner_id=mail_gateway_bot.id,
+    )
+    if len(orphan_attachments) == 0:
+        return
+
+    print("")
+    print(f"Found {len(orphan_attachments)} email gateway attachments to reattach")
+    for attachment in orphan_attachments:
+        # We look for the message posted by "Internal" at the same
+        # time, in the same realm, which has a link to the attachment
+        # but no "has_attachments".  There are potentially other,
+        # later, messages (possibly from other users, to other
+        # places!) which tried to link to the attachment; we do not
+        # fix those references, because finding them efficiently is
+        # quite hard.
+        print(
+            f"Looking for a message to attach {attachment.path_id}, created {attachment.create_time}"
+        )
+        possible_matches = []
+        for model_class in (Message, ArchivedMessage):
+            possible_matches.extend(
+                model_class.objects.filter(
+                    has_attachment=False,
+                    realm_id=attachment.realm_id,
+                    sending_client_id=internal_client.id,
+                    date_sent__gte=attachment.create_time,
+                    date_sent__lte=attachment.create_time + timedelta(minutes=5),
+                    content__contains="/user_uploads/" + attachment.path_id,
```

Comment:
We can include `sender=mail_gateway_bot`, right?

Reply:
No, because these are cases where the sender was set as some other user, which caused the whole auth problem in the first place.

Reply:
Ahh, right.